### PR TITLE
packet: handle ID encoding/decoding in the Packet enums for each state/direction combination.

### PR DESCRIPTION
**Pull request by eddyb**
This is the design I discussed with @toqueteos on IRC.
It feels more idiomatic to me and it doesn't require the multiple wrapping structures.
All the `packet::$state::$direction::Packet` enums implement `packet::Packet` which provides `read` and `write` methods, higher-level than `Protocol` and abstracting over compression/encryption.
The ID encoding/decoding is done as part of those enums' `Protocol` implementation.

### CODE REVIEW

**Comment by eddyb** [comment]
See toqueteos/hematite_server#1 for slightly more changes - including usage examples.

File: src/packet.rs [hunk 1]
```diff
@@ -458,16 +387,21 @@ packets! {
...
-                    fn encode(&self, mut dst: &mut Write) -> io::Result<()> {
-                        try!(<String as Protocol>::proto_encode(&self.channel, dst));
-                        try!(dst.write_all(&self.data));
+                impl Protocol for PluginMessage {
+                    type Clean = Self;
+                    fn proto_len(this: &Self) -> usize {
+                        <String as Protocol>::proto_len(&this.channel) + this.data.len()
+                    }
+                    fn proto_encode(this: &Self, mut dst: &mut Write) -> io::Result<()> {
+                        try!(<Var<i32> as Protocol>::proto_encode(&0x3f, dst));
```
Comment (eddyb) [inline src/packet.rs]: Oops, I forgot to remove this!

File: src/packet.rs [hunk 2]
```diff
@@ -458,16 +387,21 @@ packets! {
...
+                        try!(<String as Protocol>::proto_encode(&this.channel, dst));
+                        try!(dst.write_all(&this.data));
                         Ok(())
                     }
-                    fn decode(mut src: &mut Read, len: usize) -> io::Result<PluginMessage> {
+                    fn proto_decode(mut src: &mut Read) -> io::Result<PluginMessage> {
                         Ok(PluginMessage{
                             channel: try!(<String as Protocol>::proto_decode(src)),
-                            data: try!(src.read_exact(len)),
+                            data:  { let mut data = vec![]; try!(src.read_to_end(&mut data)); data },
```
Comment (toqueteos) [inline src/packet.rs]: This was the old proposed design, which is great!
But right now, it's not safe; we need `src` to be an adaptor or buffer or bad things could happen.

As @eddyb said, trivial.

File: src/packet.rs [hunk 3]
```diff
@@ -508,16 +442,21 @@ packets! {
...
-                    fn encode(&self, mut dst: &mut Write) -> io::Result<()> {
-                        try!(<String as Protocol>::proto_encode(&self.channel, dst));
-                        try!(dst.write_all(&self.data));
+                impl Protocol for PluginMessage {
+                    type Clean = Self;
+                    fn proto_len(this: &Self) -> usize {
+                        <String as Protocol>::proto_len(&this.channel) + this.data.len()
+                    }
+                    fn proto_encode(this: &Self, mut dst: &mut Write) -> io::Result<()> {
+                        try!(<Var<i32> as Protocol>::proto_encode(&0x17, dst));
```
Comment (eddyb) [inline src/packet.rs]: And this as well - both are encoding their IDs - which was necessary in one of the intermediary forms of this commit, but not anymore.

File: src/packet.rs (hunk 2 above)
Comment (eddyb) [inline src/packet.rs]: I guess I'll write the adapter to avoid any issues before we have decompression/decryption working.

File: src/packet.rs [hunk 4]
```diff
@@ -390,7 +320,6 @@ packets! {
         clientbound {
```
Comment (fenhl) [inline src/packet.rs]: Handshake packet is serverbound, I don't know how this happened.

<.<

File: src/packet.rs (hunk 2 above)
Comment (fenhl) [inline src/packet.rs]: Won't this break when reading from the TCP stream?

File: src/packet.rs [hunk 5]
```diff
@@ -508,16 +442,21 @@ packets! {
...
+                        try!(<String as Protocol>::proto_encode(&this.channel, dst));
+                        try!(dst.write_all(&this.data));
                         Ok(())
                     }
-                    fn decode(mut src: &mut Read, len: usize) -> io::Result<PluginMessage> {
+                    fn proto_decode(mut src: &mut Read) -> io::Result<PluginMessage> {
                         Ok(PluginMessage{
                             channel: try!(<String as Protocol>::proto_decode(src)),
-                            data: try!(src.read_exact(len)),
+                            data: { let mut data = vec![]; try!(src.read_to_end(&mut data)); data },
```
Comment (fenhl) [inline src/packet.rs]: Same as above: this will consume the entire TCP stream.

**Comment by fenhl** [comment]
Overall a good change; we don't need one big packet enum if we know state and direction anyway. However, packets implement `Protocol` now, which implies that they can be decoded without knowing their length. This is not true. To prevent mod authors from using them in plugin messages, we should either use a different trait for them or clearly state this exception in the docs.

File: src/packet.rs (hunk 2 above)
Comment (eddyb) [inline src/packet.rs]: An adapter would keep track of bytes read since the start of the packet, so no.

**Comment by toqueteos** [comment]
With this new approach [instead of enum bloat](https://github.com/toqueteos/hematite_server/pull/1/files?diff=split#diff-58299a44cf85052ffdb9926f9a807776R104) (we still are required to do a match on every packet read) we could use [static dispatch](https://doc.rust-lang.org/book/static-and-dynamic-dispatch.html) knowing that every packet struct satifies the same Protocol trait.

``` rust
// Inside packet.rs

// Both deal with encryption and compression
pub fn write_packet<T: Protocol>(packet: T, mut dst: &mut Write) -> io::Result<()>;
pub fn read_packet<T: Protocol>(mut src: &mut Read) -> io::Result<T>;
```

Is this possible now without using trait objects?

**EDIT:**: We could add a marker trait to packet structs and use that for `write_packet` and `read_packet` so we enforce which things can get read/written as a packet.

File: src/packet.rs (hunk 4 above)
Comment (toqueteos) [inline src/packet.rs]: My ambiguous (old) naming scheme may have contributed to that.

**Comment by fenhl** [comment]
@toqueteos: Why would we want this? For `write_packet` we can instead call `packet.write(dst)`, and for `read_packet` we need to know the state and direction, so we should always call `read` on the enum.

**Comment by toqueteos** [comment]
@fenhl State doesn't suddenly change, most of the time we are are on Play state.
I don't recall where this was said exactly but the current model is one thread for every client so we store which `state` the client is in there, not in every packet read call.

Direction (read) is **always client to server** for hematite_server and **always server to client** for hematite. There's no need to explicitly tell that info when it's bound to whatever context this library is being run.

If you check usage examples this is way better than previous approach which required 3-levels deep match statements to get the packet body struct, now we just one, which still doesn't seem necessary unless I'm understanding something very wrong.

**TLDR:** You can't filter whatever comes from `TcpStreamp`, not until you read it. Direction is always the same, State is stored somewhere else (struct field on client thread).

**Comment by fenhl** [comment]
If you're only dealing with the Play state and only with serverbound packets, why don't you just import `packet::play::serverbound::*` and use that module's `Packet::read`?

We can't magically know what the state is, it needs to be saved in some value owned by the connection thread. Perhaps that value could have a convenience method for reading a packet, but we still need to match over the current state _somewhere_.

**Comment by toqueteos** [comment]
@fenhl Then you have 3~4 conflicts because all serverbound packets define that.
So we are forced to use uber long paths on every packet read call and then an unnnecessary match with just one case.

**Comment by toqueteos** [comment]
@eddyb Please rebase with current master.

**Comment by toqueteos** [comment]
[std::io::ReadExt::take](http://doc.rust-lang.org/std/io/trait.ReadExt.html#method.take) seems like we are looking for, we have to handle EOF though.

File: src/packet.rs [hunk 6]
```diff
@@ -17,37 +17,24 @@ pub trait Protocol {
...
-    fn write(&self, dst: &mut Write) -> io::Result<()> {
-        let len = <Var<i32> as Protocol>::proto_len(&self.id()) + self.len();
-        try!(<Var<i32> as Protocol>::proto_encode(&(len as i32), dst));
-        try!(<Var<i32> as Protocol>::proto_encode(&self.id(), dst));
-        try!(self.encode(dst));
-        Ok(())
+    fn read(mut src: &mut Read) -> io::Result<Self> {
+        let _proto_len = try!(<Var<i32> as Protocol>::proto_decode(src));
+        // TODO limit the packet to `proto_len` bytes.
+        <Self as Protocol>::proto_decode(src)
```
Comment (toqueteos) [inline src/packet.rs]:
``` rust
let mut limit_src = src.take(_proto_len as u64);
<Self as Protocol>::proto_decode(&mut limit_src)
```

Would this solve our problems with `read_to_end` in custom packets?

**Comment by fenhl** [comment]
The problem with `take` is that it consumes `self`. **Edit:** nvm, turns out this works fine.
